solr/core/src/java/org/apache/solr/search/TermsQParserPlugin.java

```diff
@@ -138,8 +164,128 @@ public Query parse() throws SyntaxError {
           bytesRefs[i] = term.toBytesRef();
         }
 
-        return method.makeFilter(fname, bytesRefs);
+        return method.makeFilter(fname, bytesRefs, localParams);
       }
     };
   }
+
+  private static abstract class TopLevelTwoPhaseIterator extends TwoPhaseIterator {
+    private final int docBase;
+    private final int nextDocBase;
+    public TopLevelTwoPhaseIterator(DocIdSetIterator approximation, int docBase, int nextDocBase) {
+      super(new PerSegmentViewDocIdSetIterator(approximation, docBase, nextDocBase));
+
+      this.docBase = docBase;
+      this.nextDocBase = nextDocBase;
+    }
+  }
+
+  private static class TopLevelDocValuesTermsQuery extends DocValuesTermsQuery implements ExtendedQuery{
+    private final String fieldName;
+    private SortedSetDocValues values;
+    private LongBitSet queryTermOrdinals;
+    private boolean matchesAtLeastOneTerm = false;
+    private boolean cache = true;
+    private boolean cacheSeparately = false;
+    private int cost;
+
+
+    public TopLevelDocValuesTermsQuery(String field, BytesRef... terms) {
+      super(field, terms);
+      this.fieldName = field;
+    }
+
+    public Weight createWeight(IndexSearcher searcher, final ScoreMode scoreMode, float boost) throws IOException {
+      values = DocValues.getSortedSet(((SolrIndexSearcher)searcher).getSlowAtomicReader(), fieldName);
```

Comment:
I suggest renaming to "topDocValues".   The "top"-ness vs "segment" needs to be made more clear in all variables names here because we dance between them.

Reply:
Another naming proposal is ```global_ordinals```.

Reply:
I went with `topLevelDocValues`, as `topDocValues` might sounds like a PQueue or a list to those unfamiliar with the code.